File: legacy/spfcheck.pl
#!/usr/bin/perl

use strict;
use warnings;
use Mail::SPF;

# Relay- and Trustclients pass right through
if ( (defined($ENV{'RELAYCLIENT'})) ||
     (defined($ENV{'TRUSTCLIENT'})) ) {
  print "\n";
  exit 0;
}

# Make sure that SMTPHELOHOST is set
if (not(defined($ENV{'SMTPHELOHOST'}))) { print "\n"; exit 0; };

my $spf_server = Mail::SPF::Server->new();
my $request = Mail::SPF::Request->new(ip_address    => $ENV{'TCPREMOTEIP'},
                                      identity      => $ENV{'SMTPMAILFROM'},
                                      helo_identity => $ENV{'SMTPHELOHOST'},
				      scope         => 'mfrom');

my $result = $spf_server->process($request);

# Reject the client if SPF check returns "fail"
if ($result->code eq 'fail') {
  print STDERR getppid()." SPF check failed for $ENV{'SMTPMAILFROM'}\n";
  print STDERR getppid()." Mail recipient would have been $ENV{'SMTPRCPTTO'}\n";
  print "E451 SPF check failed: ".$result->local_explanation."\n";
} else {
  print "\n";
}

exit;

File: spfcheck.py
#!/usr/bin/python2

import os
import spf
import sys

# RELAYCLIENTs and TRUSTCLIENTs can pass
if "RELAYCLIENT" in os.environ or "TRUSTCLIENT" in os.environ:
  print
  exit()

# Get the client's IP from %ENV
remoteip = ""
if os.environ.get('TCPREMOTEIP'):
  remoteip = os.environ['TCPREMOTEIP']
if os.environ.get('TCP6REMOTEIP'):
  remoteip = os.environ['TCP6REMOTEIP']
if os.environ.get('SSLREMOTEIP'):
  remoteip = os.environ['SSLREMOTEIP']

# Exit if we can not get client IP
if len(remoteip) == 0:
  print
  exit()

# Exit if the host did not say hel(l)o
if not os.environ.get('SMTPHELOHOST'):
  print
  exit()

# Exit if there is no sender
if not os.environ.get('SMTPMAILFROM'):
  print
  exit()

# Run the SPF check
result, explanation = spf.check2(i=remoteip, s=os.environ['SMTPMAILFROM'], h=os.environ['SMTPHELOHOST'])

# Only fails will be rejected
if result == "fail":
  print >> sys.stderr, "%d SPF check failed for %s" % (os.getppid(), os.environ['SMTPMAILFROM'])
  print "E451 SPF check failed [%s]" % explanation
else:
  print

exit()
